File: javascript/cote/LTCD/중복문자_없는_가장_긴_부분_문자열.mjs

const TEST_CASES = 10
const N = 10000

function initialize() {

	const alphabets = "Z".charCodeAt(0) - "A".charCodeAt(0);
	
	function getAlpabet(n) {
		return String.fromCharCode("A".charCodeAt(0) + n % alphabets)
	}

	let s = ""
	let n = Math.floor(Math.random() * (N - 1) + 1)

	for (let i = 0; i < n; i++) {
		s += getAlpabet(Math.floor(Math.random() * alphabets))
	}

	return s
}

function solve(s) {
	const used = new Map();
	let [ret, start] = [0, 0]

	for (let i = 0; i < s.length; i++) {

		if (used.has(s[i]) && start <= used.get(s[i])) {
			start = used.get(s[i]) + 1
		} else {
			ret = Math.max(ret, i - start + 1)
		}
		used.set(s[i], i)
	}

	return ret
}


for (let it = 0; it < TEST_CASES; it++) {
	const s = initialize()
	console.log(solve(s))
}

File: python/cote/LTCD/중복문자_없는_가장_긴_부분_문자열.py
import random

TEST_CASES = 10
N = 10000


def solve(s: str) -> int:
    used = {}
    max_length = start = 0
    for i, c in enumerate(s):
        if c in used and start <= used[c]:
            start = used[c] + 1
        else:
            max_length = max(max_length, i - start + 1)
        used[c] = i

    return max_length


if __name__ == "__main__":
    random.seed(43)
    for _ in range(TEST_CASES):
        l = random.randrange(1, N)
        s = "".join([chr(random.randrange(ord('a'), ord('z') + 1)) for _ in range(l)])
        print(solve(s))
